### PR TITLE
Fix options databinding by adding elements if they are not in the initial map

### LINKED ISSUE #3538: Custom data in Type field in techreport not stored
Originally reported in the forum:
http://discourse.jabref.org/t/content-of-field-type-disappears-when-saving/947

- [x] I have tested the latest master version from http://builds.jabref.org/master/ and the problem persists

1. Create a techreport entry
2. Enter a type e.g. myCustomTypeOfReports
3. Move to other field -> Content is gone 

I suspect that it is a simliar databinding issue in relation to the date field #3483

### PR DESCRIPTION
**Pull request by Siedlerchr**
Fixes #3538 

I dunno if this is the best idea, but it works...

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Screenshots added (for bigger UI changes)
- [x] Manually tested changed features in running JabRef
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)
- [ ] If you changed the localization: Did you run `gradle localizationUpdate`?

### CODE REVIEW

**Comment by tobiasdiez** [comment]
Can you please explain the reason for the bug? In general, the string converter should not modify the source or target map.

**Comment by Siedlerchr** [comment]
The reason for the bug is that the fromString method does a lookup in the map with the current txt in the field. If you enter a non predefined String into it, the map lookup returns null because not found.

File: src/main/java/org/jabref/gui/fieldeditors/MapBasedEditorViewModel.java [hunk 1]
```diff
@@ -38,7 +39,11 @@ public T fromString(String string) {
                 if (string == null) {
                     return null;
                 } else {
-                    return getItemMap().getOrDefault(string, null);
+
+                    if (getItemMap().get(string) == null) {
+                        getItemMap().put(string, (T) string);
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/fieldeditors/MapBasedEditorViewModel.java]: Ok thanks!
According to the documentation
> a string converter property is provided to allow for developers to specify how to translate a users string into an object of type T, such that the value property may contain it. By default the converter simply returns the String input as the user typed it, which therefore assumes that the type of the editable ComboBox is String.

So you are right that the string converter should return something non-null even when the text is not recognized. However, there is no need to add it to the item map. Moreover, we need to handle the case when `T` is not a string type. Thus I would propose to create a new protected method that converts an unknown string to a `T` object. If not overridden, it should try to cast to `T` as in your code above (and throw an exception with details how to properly implement it if casting was not successful).

Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/fieldeditors/MapBasedEditorViewModel.java]: I now added a method for checking.
We also need such kind for dates to allow custom values which we currently don't support. All input in the field will be cleared if not parseable

**Pull request merged**